### DISCUSSION TITLE
Issues with adaptivetimestep

### DISCUSSION THREAD

**Discussion by sidharthsarmah** [idaholab/moose#24962] [Q&A General]
Hi @GiudGiud ,

I am trying to compare my results with that obtain from ABAQUS, I got two issues:

I tired to solve the problem in MOOSE but it does not converge.
```
25 Nonlinear |R| = 8.920182e-05
Nonlinear solve did not converge due to DIVERGED_MAX_IT iterations 25
 Solve Did NOT Converge!
Currently Executing
  Finished Solving                                                                       [ 24.18 s] [  289 MB]
Aborting as solve did not converge

Solve failed, cutting timestep.


```

```
[Executioner]

  type = Transient
  solve_type = 'NEWTON'

  l_tol = 1e-10
  nl_max_its = 25
  nl_rel_tol = 1e-11
  nl_abs_tol = 1e-10
  l_max_its = 10

  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
 
  start_time = 0.0
  end_time = 118
 dtmax = 1.0

  [TimeStepper]
    type = IterationAdaptiveDT
    optimal_iterations = 20
    iteration_window = 9
    growth_factor = 2
    cutback_factor = 0.5
    dt = 0.01
  []

  line_search = 'none'
  automatic_scaling = true

[]
```

<img width="436" alt="image" src="https://github.com/idaholab/moose/assets/110927986/10dcafe5-46d5-481e-aaf0-2f19f3dd856c">

**Comment by GiudGiud** [comment]
Follow the advice on this page for the non-convergence:
https://mooseframework.inl.gov/help/troubleshooting.html

For the time steps, you can now compose time steppers to create any time evolution you like (needs to converge though!)
https://mooseframework.inl.gov/syntax/Executioner/TimeStepper/index.html

**Reply by sidharthsarmah** [reply]
Hi @GiudGiud,

I figured out the issue with the convergence, its related to the load.

See the loading figure, I tried changing the magnitude of load (P) with a constant. The real constant is 10.4 but I tried for 0.1 and 1 first.
<img width="474" alt="image" src="https://github.com/idaholab/moose/assets/110927986/611cf88d-11dd-46f9-ab40-4ffecf038177">
The results matches for 0.1 but not for 1.0, with ABAQUS, is there any restriction of load that we can apply in MOOSE?
<img width="752" alt="image" src="https://github.com/idaholab/moose/assets/110927986/c92f4795-6109-4915-97ad-4881d517e17d">

**Reply by GiudGiud** [reply]
I dont think so
For 1.0, does MOOSE not converge?

**Reply by sidharthsarmah** [reply]
Hi @GiudGiud, no its not converging.

```
25 Nonlinear |R| = 1.561106e-04
Nonlinear solve did not converge due to DIVERGED_MAX_IT iterations 25
 Solve Did NOT Converge!
Currently Executing
  Finished Solving                                                                       [ 24.10 s] [  278 MB]
Aborting as solve did not converge

Solve failed, cutting timestep.
```

```
[Executioner]

  type = Transient
  solve_type = 'NEWTON'

  l_tol = 1e-10
  nl_max_its = 25
  nl_rel_tol = 1e-11
  nl_abs_tol = 1e-10
  l_max_its = 10

  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
 
  start_time = 0.0
  end_time = 118
  dtmax = 1.0

  [TimeStepper]
    type = IterationAdaptiveDT
    growth_factor = 1.5
    cutback_factor = 0.5
    dt = 0.01
  []

  line_search = 'none'
  automatic_scaling = true

[]
```

**Reply by GiudGiud** [reply]
can you attach the full log

**Reply by sidharthsarmah** [reply]
hi @GiudGiud,

I tried again to check, it works till `0.2 X load` but not beyond it.

Here it is :
```
*** ERROR ***
The following error occurred in the object "TimeStepper", of type "IterationAdaptiveDT".

Solve failed and timestep already at dtmin, cannot continue!

Stack frames: 11
0: 0   libmesh_opt.0.dylib                 0x0000000104a90264 libMesh::print_trace(std::__1::basic_ostream<char, std::__1::char_traits<char>>&) + 996
1: 1   libmoose-opt.0.dylib                0x0000000103fb3408 moose::internal::mooseErrorRaw(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>) + 676
2: 2   libmoose-opt.0.dylib                0x0000000103f2fb54 callMooseErrorRaw(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&, MooseApp*) + 100
3: 3   libmoose-opt.0.dylib                0x00000001037ce080 void MooseObject::mooseError<char const (&) [61]>(char const (&) [61]) const + 328
4: 4   libmoose-opt.0.dylib                0x0000000103a9a1ec IterationAdaptiveDT::computeFailedDT() + 492
5: 5   libmoose-opt.0.dylib                0x0000000103aa3ccc TimeStepper::computeStep() + 152
6: 6   libmoose-opt.0.dylib                0x00000001033ef684 Transient::execute() + 120
7: 7   libmoose-opt.0.dylib                0x0000000103f8b0d8 MooseApp::executeExecutioner() + 268
8: 8   libmoose-opt.0.dylib                0x0000000103f8755c MooseApp::run() + 1468
9: 9   tensor_mechanics-opt                0x00000001006cbb40 main + 160
10: 10  dyld                                0x000000019dbe7f28 start + 2236
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor
(moose) sidharthsarmah@Sidharths-MacBook-Air tensor_mechanics % 
```

**Reply by GiudGiud** [reply]
no that's not it.  That s just the end of it
Please get the full log and attach (not paste because it s real long) here

**Reply by GiudGiud** [reply]
docx is too unsafe for us :( could you please do .txt for the format

**Reply by sidharthsarmah** [reply]
Here it is @GiudGiud 
[moose_error.txt](https://github.com/idaholab/moose/files/12088995/moose_error.txt)

**Reply by GiudGiud** [reply]
is something happening at time 82-83s ?

**Reply by sidharthsarmah** [reply]
Hi @GiudGiud , No its just a continuous load that I applied for 118 s, but it does not occur, when load magnitude is increased.

**Reply by GiudGiud** [reply]
Weird. I would say restart at the time step right before it crashes the use SVD preconditiioning to study why the nonlinear is not being solved

**Reply by sidharthsarmah** [reply]
Hi @GiudGiud, you mean I should load till 82 s (when it fails to converge) to check what happens?

**Reply by GiudGiud** [reply]
load until then, make a checkpoint there, then load from there

if you turned on "checkpoint = true" in your previous simulation you should already have a checkpoint file

**Reply by sidharthsarmah** [reply]
Hi @GiudGiud , I tried it for multiplying by 10.4 factor, which is the real condition, the convergence issue arise at 2.4 s.

<img width="568" alt="image" src="https://github.com/idaholab/moose/assets/110927986/9c484878-4cbd-40bc-944b-852985f98c83">

Can you suggest how to proceed?

<img width="634" alt="image" src="https://github.com/idaholab/moose/assets/110927986/a9b0384b-7216-4df5-b2b0-4e63d105e691">

**Reply by GiudGiud** [reply]
did you try SVD after restarting from that time step?
see some advice here
https://mooseframework.inl.gov/moose/application_usage/failed_solves.html

**Reply by sidharthsarmah** [reply]
Hi @GiudGiud , I am a bit confused about how to proceed.

I tried shifting the condition to elastic, where I can still see converge issue, but after a longer time and load (N), is there any issue with the point load, maybe making the stress infinite after sometime which the structure cannot sustain?

```
***Plastic***
[Materials]
  [./elasticity_tensor]
    type = ADComputeIsotropicElasticityTensor
    youngs_modulus = 211000
    poissons_ratio = 0.3
  [../]

  [./isotropic_plasticity]
    type = ADIsotropicPlasticityStressUpdate
    yield_stress = 1220
    hardening_function = 'hard_function'
  [../]

  [./radial_return_stress]
    type = ADComputeMultipleInelasticStress
    inelastic_models = 'isotropic_plasticity'
  [../]

[]
```

```
***Elastic***
[Materials]
  [./elasticity_tensor]
    type = ADComputeIsotropicElasticityTensor
    youngs_modulus = 211000
    poissons_ratio = 0.3
  [../]

[./stress]
  type = ADComputeLinearElasticStress
[../]

[]
```
![image](https://github.com/idaholab/moose/assets/110927986/bf4062c7-e1e5-48fb-b3a5-860b15612542)

**Reply by GiudGiud** [reply]
I don’t know either this isn’t my field of expertise. It seems odd that things stop converging at a given time without a physical cause. It seems like something is happening at that time. 

You can try working with the scaling of the equations. You can also reduce the linear tolerance, and use LU with a nonzero factor shift in case you have a zero on the diagonal at that step

**Comment by sidharthsarmah** [comment]
Hi @GiudGiud,

I think this is the problem related with this problem.

I ran the problem till 22 s where there was no issue with convergence. 

Till then the `VonMises stress = 7200 MPa` in 22 s, and the material `Yield strength = 1220 MPa`, it means there will be ample plastic deformation till 118s, but I defined the `strain=small`.

<img width="731" alt="image" src="https://github.com/idaholab/moose/assets/110927986/50e5ca34-914e-406c-b060-7eb99d706865">

But when I use `strain=finite`, there is no convergence and this is very confusing to me, can you explain why I cannot use `finite` deformation because it falls in the plastic deformation.

```
[Modules/TensorMechanics/Master]

  [./all]
    strain = finite
    incremental = true
    add_variables = true
    use_automatic_differentiation = true
    save_in = 'force_x force_y'
  [../]

[]
```

**Reply by GiudGiud** [reply]
@hugary1995

**Reply by sidharthsarmah** [reply]
Hi @GiudGiud , 

I was able to solve it, please check it if it's correct, the load applied is very high and results in unrealistic deformation, but it can be solved. I have made this changes.

```
[GlobalParams]
  displacements = 'disp_x disp_y'
[]

[Modules/TensorMechanics/Master]
  [./all]
    strain = finite
    incremental = true
    add_variables = true
    use_automatic_differentiation = true
    save_in = 'force_x force_y'
  [../]
[]

[Executioner]
  type = Transient
  solve_type = 'PJFNK'
  petsc_options = '-snes_ksp_ew'
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
  nl_abs_tol = 1e-10
  l_max_its = 20
  start_time = 0.0
  end_time = 118
  dtmax = 1.0
  [TimeStepper]
    type = IterationAdaptiveDT
    growth_factor = 1.5
    cutback_factor = 0.5
    dt = 0.1
  []
[]
```
<img width="360" alt="image" src="https://github.com/idaholab/moose/assets/110927986/a4feb1cf-ce57-47da-bff8-811556dc72d3">

**Reply by GiudGiud** [reply]
To check if it s correct you should lead a mesh & time grid convergence study. 

@recuero are these parameters good for this kind of loading tests?

**Reply by recuero** [reply]
@sidharthsarmah what parameters did you change to make it work?

For large deformation problems, PJFNK without AD or Newton with AD is recommended. If you switched from Newton to PJFNK, that may have helped obtain improved convergence.

**Reply by sidharthsarmah** [reply]
@recuero , just changed NEWTON to PJFNK with AD, it still worked.

**Reply by recuero** [reply]
Yes, but it's probably more efficient with NEWTON.

**Reply by sidharthsarmah** [reply]
hi @GiudGiud ,

Instead of `make -j 4` for building, I typed `make -j 5` by mistake, now I am getting this error, please suggest.
```
*** ERROR ***
A 'ADCZMRealVectorScalar' is not a registered object.

If you are trying to find this object in a dynamically linked library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.

```

**Reply by chunhuizhao478** [reply]
I realize today that the `ADCZMRealVectorScalar` is not present in the MOOSE code, is this not ready? @hugary1995 @GiudGiud , thanks!

**Reply by GiudGiud** [reply]
What's the context? Is this in a PR already?

**Reply by sidharthsarmah** [reply]
Hi @GiudGiud , I want a time step which is 1000 s for time 86400 s, and 0.35 s after it reaches 86400, something like this:

<img width="571" alt="image" src="https://github.com/idaholab/moose/assets/110927986/04245cb2-8047-401c-b4b2-c7d028050499">

```
[./dts]
    type = PiecewiseLinear
    x = '0   86400 86638'
    y = '1000 0.35 0.35'
  [../]

[Executioner]

  type = Transient

  solve_type = 'NEWTON'
  line_search = none

  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'

  automatic_scaling = true
  l_tol = 1e-10
  nl_rel_tol = 1e-10
  nl_abs_tol = 1e-10
  nl_max_its = 10
  start_time = 0.0
  end_time = 86638
  dtmax = 1000
  [./TimeStepper]
    type = FunctionDT
    function = dts
  [../]
[]
```

**Reply by GiudGiud** [reply]
Hello

You may want to ease this transition. Dividing the time step by 3000 from a step to another won't do wonders for most time integration schemes. Here implicit Euler might be fine, though it seems to be struggling from what you are reporting. 

Guillaume

**Reply by sidharthsarmah** [reply]
Hi @GiudGiud , you are correct. Do you think this could be the reason, I am having this issue?

The problem I am solving includes diffusion from the boundary of a tensile specimen without any load for 24, 48 and 72 hours, and then loading it with a disp BC till failure for 3 mins alongside setting the boundary values of concentration to 0.

The simulation works perfectly for 24, 48 hours but not for 72 hours. In that case, diffusion works well for 71.99 hours and then do not converge, this is very confusion as till now its there's no loading and just diffusion to a static body, could you suggest its because of this sudden transition?

**Reply by GiudGiud** [reply]
You can test that for yourself by changing the time steps profile. Or by trying no load and no time step change for 73 hours. 
But that would be my suspicion
